### PR TITLE
Rust async unlock

### PR DESCRIPTION
**Pull request by benma**

### CODE REVIEW

**Comment by benma** [comment]
Test by normal unlock of the device - the async unlock is blocked on for now (will be made properly async with cancel later).

**Comment by benma** [comment]
unlock.rs will eventually replace:

https://github.com/digitalbitbox/bitbox02-firmware/blob/master/src/workflow/unlock.c
https://github.com/digitalbitbox/bitbox02-firmware/blob/master/src/workflow/get_mnemonic_passphrase.c
https://github.com/digitalbitbox/bitbox02-firmware/blob/master/src/workflow/unlock_bip39.c
https://github.com/digitalbitbox/bitbox02-firmware/blob/master/src/ui/workflow_stack.c

which all exist to make async work.

File: src/workflow/unlock.c [hunk 1]
```diff
@@ -183,20 +184,7 @@ workflow_t* workflow_unlock(void (*callback)(bool result, void* param), void* ca
...
-{
-    bool* result_ptr = (bool*)param;
-    *result_ptr = result;
-    workflow_blocking_unblock();
-}
-
 bool workflow_unlock_blocking(void)
 {
-    bool result = false;
-    workflow_stack_start_workflow(workflow_unlock(_workflow_unlock_completed, &result));
```
[the remark this answers is not in the document]
Comment (benma) [reply src/workflow/unlock.c:199]: yeah, the (old-style) async unlock is currently in in U2F (u2f.c). I can remove it and the other files as soon as I use the new-style async in U2F. I plan on doing that mostly at the same time as using it in HWW (the normal hardware wallet usb processing).

File: src/rust/bitbox02/src/ui.rs [hunk 2]
```diff
@@ -204,3 +204,33 @@ pub fn screen_process() {
...
+        let callback = &mut *callback_ptr;
+        callback();
+    }
+
+    let component = unsafe {
+        bitbox02_sys::status_create(
+            crate::str_to_cstr_force!(text, 200).as_ptr(),
+            status_success,
+            Some(c_callback::<F>),
+            // TODO: from_raw
```
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/bitbox02/src/ui.rs:227]: Next PR is fixing the TODO with explanations. Short story: the callback is moved to C land, so it's a memory leak in Rust. from_raw brings it back and drops it properly.

File: src/rust/bitbox02/src/ui.rs [hunk 3]
```diff
@@ -204,3 +204,33 @@ pub fn screen_process() {
...
+        F2: FnMut(),
+    {
+        let callback_ptr = param as *mut F2;
+        let callback = &mut *callback_ptr;
+        callback();
+    }
+
+    let component = unsafe {
+        bitbox02_sys::status_create(
+            crate::str_to_cstr_force!(text, 200).as_ptr(),
```
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/bitbox02/src/ui.rs]: Good catch, it was not even an off by one error, I just guessed some safe upper bound :innocent: but yeah makes sense to have it the same.

I couldn't figure out how to use a Rust const and use it in a macro call, gave me a compiler error :confused:

File: src/rust/bitbox02/src/keystore.rs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+pub fn unlock(password: &Password) -> Result<(), Error> {
+    let mut remaining_attempts: u8 = 0;
+    match unsafe { bitbox02_sys::keystore_unlock(password.as_cstr(), &mut remaining_attempts) } {
+        keystore_error_t::KEYSTORE_OK => Ok(()),
+        keystore_error_t::KEYSTORE_ERR_INCORRECT_PASSWORD => {
+            Err(Error::IncorrectPassword { remaining_attempts })
+        }
+        keystore_error_t::KEYSTORE_ERR_MAX_ATTEMPTS_EXCEEDED => Err(Error::Unknown),
```
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/bitbox02/src/keystore.rs:36]: Imho we could rather remove that entry form keystore.c, as the unit test actually covers the reset case by checking that reset_reset() was called. Checking the enum entry in addition to that seems redundant.

File: src/rust/bitbox02/src/ui.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/bitbox02/src/ui.rs:227]: Like so: https://github.com/digitalbitbox/bitbox02-firmware/pull/440

File: src/rust/bitbox02/src/ui.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (NickeZ) [reply src/rust/bitbox02/src/ui.rs]: That's beause you parse it as a `literal` in the macro: `$len:literal` try expression instead `expr`.

File: src/rust/bitbox02/src/ui.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (x1ddos) [reply src/rust/bitbox02/src/ui.rs:227]: Nice. Yeah, I just meant the TODO belongs to c_callback, somewhere on line 219.
But never mind! It's gone with #440. :)

**Pull request merged**